### ISSUE TITLE
ClangCL simdjson 15% slower than Clang on WSL

### ISSUE THREAD

**Issue by jkeiser** [simdjson/simdjson#848]
| Platform | File | Overall | Stage 1 | Stage 2 |
|---|---|---|---|---|
| ClangCL 9.0.0 | twitter.json | 2.2221 | 5.4161 | 4.6401 |
| WSL2 clang 9.0.0 | twitter.json | 2.5829 | 5.7099 | 4.7375 |
| ClangCL 9.0.0 | twitter10.json | 2.1904 | 4.8541 | 4.1482 |
| WSL2 clang 9.0.0 | twitter10.json | 2.2023 | 4.3864 | 4.0769 |

It looks like the difference is mostly in stage 1, and isn't all that huge, but it's still puzzling.

Methodology:
* ClangCL: `git clean -ffdx build && cmake -B build -T ClangCL && cmake --build build --target parse --config Release && build\benchmark\Release\parse jsonexamples\twitter.json`
* WSL2 clang 9.0.0: `cd ~/simdjson && git clean -ffdx build && mkdir build && cd build && CXX=clang++-9 CC=clang-9 cmake .. && cmake --build . --target parse && benchmark/parse ../jsonexamples/twitter.json`

**Comment by lemire** [comment]
Todo: try with files having different size (tiny all the way to really large) to see whether the problem is related to memory accesses.

**Comment by jkeiser** [comment]
@lemire I made a file that is basically twitter.json inside an array 10 times. ClangCL and WSL Clang perform very similarly on the larger file.

The source of the issue is pretty clear. Allocation is less than O(n) complexity--in fact, almost constant time, so maybe it's O(log n) or something:

```
λ build\benchmark\Release\parse jsonexamples\twitter.json jsonexamples/twitter10.json
number of iterations 200

jsonexamples\twitter.json
=========================
     9867 blocks -     631515 bytes - 55263 structurals (  8.8 %)
special blocks with: utf8      2284 ( 23.1 %) - escape       598 (  6.1 %) - 0 structurals      1287 ( 13.0 %) - 1+ structurals      8581 ( 87.0 %) - 8+ structurals      3272 ( 33.2 %) - 16+ structurals         0 (  0.0 %)
special block flips: utf8      1104 ( 11.2 %) - escape       642 (  6.5 %) - 0 structurals       940 (  9.5 %) - 1+ structurals       940 (  9.5 %) - 8+ structurals      2593 ( 26.3 %) - 16+ structurals         0 (  0.0 %)

All Stages
|    Speed        :  28.6178 ns per block ( 81.63%) -   0.4472 ns per byte -   5.1101 ns per structural -   2.2362 GB/s
|- Allocation
|    Speed        :   2.5132 ns per block (  7.17%) -   0.0393 ns per byte -   0.4488 ns per structural -  25.4643 GB/s
|- Stage 1
|    Speed        :  12.3227 ns per block ( 35.15%) -   0.1926 ns per byte -   2.2004 ns per structural -   5.1934 GB/s
|- Stage 2
|    Speed        :  13.5184 ns per block ( 38.56%) -   0.2112 ns per byte -   2.4139 ns per structural -   4.7340 GB/s

3541.1 documents parsed per second (best)

jsonexamples/twitter10.json
===========================
    98674 blocks -    6315172 bytes - 552641 structurals (  8.8 %)
special blocks with: utf8     22768 ( 23.1 %) - escape      5932 (  6.0 %) - 0 structurals     12809 ( 13.0 %) - 1+ structurals     85866 ( 87.0 %) - 8+ structurals     32534 ( 33.0 %) - 16+ structurals         0 (  0.0 %)
special block flips: utf8     11040 ( 11.2 %) - escape      6444 (  6.5 %) - 0 structurals      9464 (  9.6 %) - 1+ structurals      9464 (  9.6 %) - 8+ structurals     25364 ( 25.7 %) - 16+ structurals         0 (  0.0 %)

All Stages
|    Speed        :  29.6539 ns per block ( 86.29%) -   0.4633 ns per byte -   5.2948 ns per structural -   2.1582 GB/s
|- Allocation
|    Speed        :   0.3588 ns per block (  1.04%) -   0.0056 ns per byte -   0.0641 ns per structural - 178.3947 GB/s
|- Stage 1
|    Speed        :  13.8394 ns per block ( 40.27%) -   0.2162 ns per byte -   2.4710 ns per structural -   4.6245 GB/s
|- Stage 2
|    Speed        :  15.3240 ns per block ( 44.59%) -   0.2394 ns per byte -   2.7361 ns per structural -   4.1764 GB/s

341.8 documents parsed per second (best)
```

**Comment by lemire** [comment]
>  I made a file that is basically twitter.json inside an array 10 times. 


Me too:
```bash
echo "[">twitter10.json ; for i in {1..9} ; do cat twitter.json >> twitter10.json ; echo "," >> twitter10.json ; done && cat twitter.json >> twitter10.json ; echo "]" >> twitter10.json
```

> The source of the issue is pretty clear.

 Is it?

Meanwhile... here are the Linux results...

```
$ ./benchmark/parse ../jsonexamples/twitter.json
number of iterations 200

../jsonexamples/twitter.json
============================
     9867 blocks -     631515 bytes - 55263 structurals (  8.8 %)
special blocks with: utf8      2284 ( 23.1 %) - escape       598 (  6.1 %) - 0 structurals      1287 ( 13.0 %) - 1+ structurals      8581 ( 87.0 %) - 8+ structurals      3272 ( 33.2 %) - 16+ structurals         0 (  0.0 %)
special block flips: utf8      1104 ( 11.2 %) - escape       642 (  6.5 %) - 0 structurals       940 (  9.5 %) - 1+ structurals       940 (  9.5 %) - 8+ structurals      2593 ( 26.3 %) - 16+ structurals         0 (  0.0 %)

All Stages
|    Speed        :  26.3991 ns per block ( 99.23%) -   0.4125 ns per byte -   4.7139 ns per structural -  2.42 GB/s -  2.26 GiB/s
|    Cycles       :  97.5280 per block    ( 99.29%) -   1.5240 per byte    -  17.4150 per structural    -    3.694 GHz est. frequency
|    Instructions : 315.2109 per block    (100.00%) -   4.9255 per byte    -  56.2854 per structural    -    3.232 per cycle
|    Misses       :    2763 branch misses ( 92.63%) - 0 cache misses (  0.00%) - 78967.00 cache references
|- Stage 1
|    Speed        :  12.1005 ns per block ( 45.48%) -   0.1891 ns per byte -   2.1607 ns per structural -  5.29 GB/s -  4.93 GiB/s
|    Cycles       :  44.6924 per block    ( 45.50%) -   0.6984 per byte    -   7.9805 per structural    -    3.693 GHz est. frequency
|    Instructions : 156.9514 per block    ( 49.79%) -   2.4525 per byte    -  28.0259 per structural    -    3.512 per cycle
|    Misses       :    1225 branch misses ( 41.07%) - 0 cache misses (  0.00%) - 27415.00 cache references
|- Stage 2
|    Speed        :  14.2594 ns per block ( 53.60%) -   0.2228 ns per byte -   2.5462 ns per structural -  4.49 GB/s -  4.18 GiB/s
|    Cycles       :  52.6771 per block    ( 53.63%) -   0.8231 per byte    -   9.4063 per structural    -    3.694 GHz est. frequency
|    Instructions : 158.0996 per block    ( 50.16%) -   2.4705 per byte    -  28.2310 per structural    -    3.001 per cycle
|    Misses       :    1520 branch misses ( 50.96%) - 0 cache misses (  0.00%) - 51602.00 cache references

3838.7 documents parsed per second (best)
dlemire@skylake:~/CVS/github/simdjson/build$ ./benchmark/parse ../jsonexamples/twitter10.json
number of iterations 200

../jsonexamples/twitter10.json
==============================
    98674 blocks -    6315172 bytes - 552641 structurals (  8.8 %)
special blocks with: utf8     22768 ( 23.1 %) - escape      5932 (  6.0 %) - 0 structurals     12809 ( 13.0 %) - 1+ structurals     85866 ( 87.0 %) - 8+ structurals     32534 ( 33.0 %) - 16+ structurals         0 (  0.0 %)
special block flips: utf8     11040 ( 11.2 %) - escape      6444 (  6.5 %) - 0 structurals      9464 (  9.6 %) - 1+ structurals      9464 (  9.6 %) - 8+ structurals     25364 ( 25.7 %) - 16+ structurals         0 (  0.0 %)

All Stages
|    Speed        :  29.0085 ns per block ( 99.61%) -   0.4533 ns per byte -   5.1795 ns per structural -  2.21 GB/s -  2.05 GiB/s
|    Cycles       : 106.9613 per block    ( 99.70%) -   1.6713 per byte    -  19.0981 per structural    -    3.687 GHz est. frequency
|    Instructions : 314.8647 per block    (100.00%) -   4.9198 per byte    -  56.2196 per structural    -    2.944 per cycle
|    Misses       :   34690 branch misses ( 99.25%) - 415106 cache misses ( 99.68%) - 822777.00 cache references
|- Stage 1
|    Speed        :  13.2878 ns per block ( 45.63%) -   0.2076 ns per byte -   2.3726 ns per structural -  4.82 GB/s -  4.49 GiB/s
|    Cycles       :  49.0549 per block    ( 45.72%) -   0.7665 per byte    -   8.7588 per structural    -    3.692 GHz est. frequency
|    Instructions : 156.8025 per block    ( 49.80%) -   2.4501 per byte    -  27.9974 per structural    -    3.196 per cycle
|    Misses       :   19754 branch misses ( 56.52%) - 144381 cache misses ( 34.67%) - 284528.00 cache references
|- Stage 2
|    Speed        :  15.6154 ns per block ( 53.62%) -   0.2440 ns per byte -   2.7882 ns per structural -  4.10 GB/s -  3.82 GiB/s
|    Cycles       :  57.6463 per block    ( 53.73%) -   0.9007 per byte    -  10.2928 per structural    -    3.692 GHz est. frequency
|    Instructions : 158.0453 per block    ( 50.19%) -   2.4695 per byte    -  28.2193 per structural    -    2.742 per cycle
|    Misses       :   14743 branch misses ( 42.18%) - 263016 cache misses ( 63.16%) - 536564.00 cache references

```

So a 10% difference on the small file... which according to profiling (`perf record`)  can be entirely accounted by the price we pay due to memcpy.

So it seems like we are calling, at some point, a copy.

Where?


Could you run both under [vtunes](https://software.intel.com/en-us/vtune) or whatever the closest equivalent to `perf` is under Windows? If we are bounded on large files by a copy under Linux, the same might be true under Windows. But maybe the copy is somehow more expensive under Windows.

**Comment by lemire** [comment]
> So it seems like we are calling, at some point, a copy. (...) Where?

I create a large numbers file...

```
echo "[">numbers100.json ; for i in {1..99} ; do cat numbers.json >> numbers100.json ; echo "," >> numbers100.json ; done && cat numbers.json >> numbers100.json ; echo "]" >> numbers100.json
```

I get a 4% slowdown which is, again, entirely accounted by the memcpy calls according to `perf record`.

**Comment by jkeiser** [comment]
It'll be a bit while I learn vtunes, but yeah :)

**Comment by jkeiser** [comment]
More numbers, courtesy of benchfeatures (#857):

**TLDR:** ClangCL is a lot worse when there are more structurals per block.

| Stage    | Platform |     Base | 7 Struct |    UTF-8 |   Escape |  15 Str. | 16+ Str. |   7 Struct Miss |      UTF-8 Miss |     Escape Miss |    15 Str. Miss |   16+ Str. Miss |
|----------|----------|----------|----------|----------|----------|----------|----------|-----------------|-----------------|-----------------|-----------------|-----------------|
| Stage 1  | clang 9.0.0 (WSL) |     6.87 |     2.54 |     4.68 |        0 |     1.96 |     3.92 |             2.1 |            6.52 |               0 |            5.04 |            5.32 |
| Stage 1  | ClangCL 9.0.0 |     6.89 |     9.42 |     5.06 |     0.32 |     5.34 |     10.5 |            3.32 |            6.68 |           -0.68 |            5.08 |    6.56 |
| Stage 2  | clang 9.0.0 (WSL) |     0.09 |       14 |        0 |     4.78 |     16.6 |     16.8 |             0.2 |               0 |            9.96 |            0.44 |             0.2 |
| Stage 2  | ClangCL 9.0.0 |     0.35 |     25.9 |     0.04 |     4.26 |     22.4 |     33.5 |             0.2 |           -0.36 |            8.52 |            1.32 |   -0.88 |
| All      | clang 9.0.0 (WSL) |     6.99 |     16.6 |     4.78 |     4.78 |     18.6 |     20.7 |             2.3 |            6.48 |            9.92 |            5.52 |               6 |
| All      | ClangCL 9.0.0 |     10.6 |     36.2 |      8.3 |     6.04 |     29.9 |       43 |             2.9 |            4.36 |            9.52 |            6.88 |     5.8 |

* A Base file in benchfeatures is "ab" followed by 640K spaces.
* 7 Struct is `,"ab","ab",{}` repeated every 64 bytes up to 640K.
* 15 Str. is `,"ab","ab","ab","ab","ab","ab",{}` repeated every 64 bytes up to 640K.

**Comment by lemire** [comment]
These numbers come from clang 9 / ClangCL 9 in 2020, and the stage 1 and stage 2 code (and the benchmark harness) has changed a lot since. The report no longer reflects current simdjson or current toolchains, so I'm closing it. If someone sees a significant ClangCL vs. clang gap with a recent release and a recent compiler, please open a new issue with numbers.
